File: src/Java/adventofcode2015/Day02.java

package adventofcode2015;

import java.util.Arrays;

import utils.Input;

/**
 * https://adventofcode.com/2015/day/2
 */
public class Day02 {

    private static int wrap(int[] dims) {
        int a = dims[0] * dims[1], b = dims[1] * dims[2], c = dims[2] * dims[0];
        return 2 * (a+b+c) + a;
    }

    private static int ribbon(int[] dims) {
        return 2 * (dims[0] + dims[1]) + dims[0] * dims[1] * dims[2];
    }

    public static void main(String[] args) {
        int[][] dimensions = new Input(2015, 2)
                .strings().stream()
                .map(s -> Arrays.stream(s.split("x"))
                        .mapToInt(Integer::parseInt).sorted().toArray())
                .toArray(int[][]::new);

        System.err.println("paper: " +
                Arrays.stream(dimensions).mapToInt(Day02::wrap).sum());

        System.err.println("ribbon: " +
                Arrays.stream(dimensions).mapToInt(Day02::ribbon).sum());
    }

}

File: src/Python/2018/Day02.py

from utils import Input, printResult

# https://adventofcode.com/2018/day/2

lines = sorted(Input(2018, 2).lines())
a, b = 0, 0
for line in lines:
    if any(line.count(c) == 2 for c in line): a += 1
    if any(line.count(c) == 3 for c in line): b += 1
printResult(1, a * b)

for (a, b) in zip(lines[:-2], lines[1:]):
    for i in range(len(a)):
        if a[:i] + a[i+1:] == b[:i] + b[i+1:]:
            printResult(2, a[:i] + a[i+1:])
